### PR TITLE
Fail fast if a stored class won't be found during load

### PR DESCRIPTION
**Pull request by bamboo**
### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test ACC without PTS

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/ClassLoading.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+package org.gradle.internal.cc.impl.serialize
+
+
+private
+val defaultClassLoader = DefaultClassEncoder::class.java.classLoader
+
+
+internal
+fun classForName(name: String, classLoader: ClassLoader?): Class<*> =
```
Comment (mlopatkin) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/ClassLoading.kt:25]: 🤔 I'm starting to find our handling of `null` classloader quite confusing. `null` is a bootstrap classloader, but we're using it as a placeholder for the Gradle runtime classloader. I'd prefer us spelling out runtime classloader explicitly:
* Change `decodeClassLoader` to return non-`null` runtime classloader when loading an unscoped classloader.
* Forbid this function to accept a `null` classloader and get rid of substitution there.
* Introduce a `hasClassInRuntimeClassloader(name:String): Boolean` with the implementation like
```kotlin
try {
  Class.forName(name, defaultClassLoader)
  return true
} catch (e: ClassNotFoundException) {
  return false
}
```
and use it to implement `ensureClassCanBeFoundInDefaultClassLoader`.

**Review by mlopatkin — CHANGES_REQUESTED** [review]
I only have suggestions about naming/structure, the general flow looks good.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/DefaultClassEncoder.kt [hunk 2]
```diff
@@ -69,19 +76,31 @@ class DefaultClassEncoder(
         } else {
             val newId = classes.putInstance(type)
             writeSmallInt(newId)
-            writeString(type.name)
-            encodeClassLoader(type.classLoader)
+            val className = type.name
+            writeString(className)
+            val classLoader = type.classLoader
+            if (!writeClassLoaderScopeOf(classLoader) && classLoader != null) {
+                // Ensure class can be found in the default classloader since its original classloader could not be encoded.
```
[the remark this answers is not in the document]
Comment (bamboo) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/DefaultClassEncoder.kt]: Fixed.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/ClassLoading.kt (hunk 1 above)
Comment (bamboo) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/ClassLoading.kt]: I moved in that direction but there's still the `Proxy` case to consider so I'm leaving the `decodeClassLoader` change to a future PR.

**Pull request merged**